yorick.ml

```diff
@@ -0,0 +1,130 @@
+     | None -> ("unset "^k^";")::list
+   ) env []) *~ " ") ^ set ^ command
+
+let (?| ) command = match Sys.command (apply_env command) with
+  | 0 -> ()
+  | x ->
+    Printf.eprintf "'%s' exited %d. Terminating with %d\n" command x x;
+    exit x
+
+let (?|.) fmt = Printf.ksprintf (?|) fmt
+
+let (?|~) fmt = Printf.ksprintf (fun command ->
+  print_endline command;
+  ?|  command
+) fmt
+
+let (?|>) fmt = Printf.ksprintf (fun command ->
+  let buf = Buffer.create (5*80) in
+  let stdout = Unix.open_process_in command in
+  try while true do Buffer.add_channel buf stdout 1 done; ""
+  with End_of_file -> close_in stdout; Buffer.contents buf
+) fmt
+
+let (?|?) fmt = Printf.ksprintf (fun command ->
+  Sys.command (apply_env command)
+) fmt
+
+let export k v = Hashtbl.replace env k (Some v)
+
+let set opts = set := opts
+
+let unset k = Hashtbl.replace env k None
+
+let map = List.map
+
+let q = Printf.sprintf "\"%s\""
+
+let ql = map q
+
+(* Not tail recursive for "performance", please choose low values for
+   [max]. The idea is that max is always small because it's hard
+   code *)
+let split_char_bounded str ~on ~max =
+  let open String in
+  if str = "" then []
+  else if max = 1 then [str]
+  else
+    let rec loop offset tokens =
+      if tokens = max - 1
+      then [sub str offset (length str - offset)]
+      else
+        try
+          let index = index_from str offset on in
+          if index = offset then
+            ""::(loop (offset + 1) (tokens + 1))
+          else
+            let token = String.sub str offset (index - offset) in
+            token::(loop (index + 1) (tokens + 1))
+        with Not_found -> [sub str offset (length str - offset)]
+    in loop 0 0
+
+let split_char_unbounded_no_trailer str ~on =
+  let open String in
+  if str = "" then []
+  else
+    let rec loop acc offset =
+      try begin
+        let index = rindex_from str offset on in
+        if index = offset then
+          (*loop (""::acc) (index - 1) -- no_trailer modified *)
+          loop acc (index - 1)
+        else
+          let token = sub str (index + 1) (offset - index) in
+          loop (token::acc) (index - 1)
+      end
+      with Not_found -> (sub str 0 (offset + 1))::acc
+    in loop [] (length str - 1)
+
+let some = function "" -> None | x -> Some x
+let list = split_char_unbounded_no_trailer ~on:' '
+let lines = split_char_unbounded_no_trailer ~on:'\n'
+let pair s = match split_char_bounded s ~on:':' ~max:2 with
+  | []      -> ("",None)
+  | [x]     -> (x, None)
+  | x::y::_ -> (x, Some y)
+
+let getenv_default var default = try Sys.getenv var with Not_found -> default
+
+let (?$) = function
+  | "@" -> List.(ql (tl (Array.to_list Sys.argv))) *~ " "
+  | v -> match some (getenv_default v "") with
+    | Some v -> v
+    | None ->
+      Printf.eprintf "I don't know what variable '%s' means and I give up.\n" v;
+      exit 1
+
+let fuzzy_bool_of_string = function "false" -> false | _ -> true
```

Comment:
An should be much more strict. It is *very* unhelpful and obfuscating that e.g. `INSTALL=/usr/local` means `INSTALL=true`. I'd suggest [this](https://github.com/dbuenzli/bos/blob/master/src/bos_os.ml#L415-L419).